### PR TITLE
refactored and improved minimal perfect hash code (regen/mph.pl)

### PR DESCRIPTION
**Pull request by demerphq**
Some time back I wrote regen/mph.pl for @khwilliamson to use in parsing our unicode properties. I had completely forgotten this code and rediscovered it when debugging https://github.com/Perl/perl5/issues/19603 (I wrote it at a time when I was under great stress after my mom passed away).

I noticed some odd things when it ran and decided to clean it up, fix the issues I saw and generally reacquaint myself with the code. While digging into it I realized it was missing certain optimizations and ran very very slowly. The version of regen/mk_invlists.pl in blead right now takes about 11.5 minutes to run on my laptop, about 9.5 of that coming from regen/mph.pl

I fixed the major of the cause of the slowness by removing "use bignum" which had been added to make it work on 32 bit perls, and then added some optimizations and improvements. The new version of regen/mk_invlists runs in about 3.5 minutes, with about 1 minute due to mph.pl and at the same time it produces a significantly (~20%) smaller blob for the keys.

Some context: mph.pl constructs a two level minimal perfect hash, and it constructs a minimal covering string of split keys. The two level minimal perfect hash algorithm involves building a special type of hash tale with exactly the number of buckets as there are keys.  The minimal covering set of split keys is a way of compressing the keys such that they are represented as a prefix and a suffix. For our use case (And many use cases) it turns out that by splitting the keys we can save a lot of space by reusing the prefixes and suffixes or parts of them, for multiple keys. 

This PR includes several improvements to both algorithms. 

The improvements to the minimal perfect hash construction means it should execute in nearly O(N) time, and the improvements to the minimal covering string algorithm mean it produces a smaller buffer to represent the keys. 

The algorithm for the minimal covering string algorithm "_squeeze" is by a former colleague of mine Ilya Sashcheka who I have unfortunately lost contact with and no longer have an email for, but I have added him to the AUTHORS file none the less. He gave me permission to use this code back when we worked together but it just never made it into core until now. 

The combination of the two algorithms means we can represent the entire data set in about 65% of the size of the raw keys themselves. This is substantially smaller than the size of code we would produce if we used Devel::Tokenizer:C which would be much larger than the 104237 bytes of raw keys we need to represent.

The net result of this patch is that it takes about 3.5 minutes to run regen/mk_invlists.pl and it produces a key buffer that is 8635 verus the current 10617 bytes. IOW, it saves about 7 minutes and 2k of ram.  As a bonus this version of the code is easier to work on and maintain as I have converted it to use OO for the public API, and has way more comments about how it works.

Included in this PR is a fix for t/op/magic.t that i noticed while working on the PR.

### CODE REVIEW

File: regen/mph.pl [hunk 1]
```diff
@@ -32,160 +112,382 @@ package MinimalPerfectHash;
...
+    # solution for the hard cases. The buckets can be divided into three
+    # levels of difficulty to solve "hard", "medium" and "trivial".
+    #
+    # * Hard buckets have more than one item in them.
+    # * Medium buckets have one item whose hash is above $MAX_SEED2.
+    # * Trivial buckets have one item whose hash is not above $MAX_SEED2.
+    #
+    # Each type of bucket uses a different algorithm to solve. Note that
+    # a "classical" two level hash would only have "hard" and "trivial"
+    # buckets, but since we supporint having a larger hash value than we
```
Comment (haarg) [inline regen/mph.pl]: typo

Comment (demerphq) [reply regen/mph.pl]: Fixed.

**Review by khwilliamson — CHANGES_REQUESTED** [review]
LGTM, but it needs to regen mk_invlist to not fail porting tests.  The CI won' run without that fix.  I did it on my box locally, and all tests passed.

**Comment by demerphq** [comment]
I modified the old last commit to include the mk_invlist.pl changes, and I added an additional sanity check patch that validates the split keys buffer and regenerated again.

**Pull request merged**